Add UniLoRA tuner to PEFT

KaiyangLi1992 opened the pull request:
## Motivation
This PR adds **UniLoRA**, a LoRA-style parameter-efficient fine-tuning method
that introduces a unified parameterization for low-rank adaptations, enabling
further reductions in the number of trainable parameters while preserving
the standard PEFT workflow.

## What's included
- UniLoRA tuner implementation
- Configuration class and registry integration
- Save/load support
- Unit tests


## Checklist
- [x] Compatible with existing PEFT APIs
- [x] Backward-compatible
- [x] Tests added

KaiyangLi1992 committed:
Add UniLoRA tuner to PEFT

githubnemo commented on src/peft/tuners/unilora/__init__.py:
```diff
@@ -0,0 +1,23 @@
+# Copyright 2024-present the HuggingFace Inc. team.
```
```suggestion
# Copyright 2025-present the HuggingFace Inc. team.
```

githubnemo commented on src/peft/tuners/unilora/layer.py:
```diff
@@ -0,0 +1,244 @@
+        if cast_to_fp32:
+            unilora_theta_d = unilora_theta_d.float()
+
+        # Changed: Replaced 'logits' with 'indices' and 'norm' with 'scales'
```
Is this a LLM comment? Please clean up after your tools :)

githubnemo reviewed (commented):
Hey @KaiyangLi1992, thanks for the PR :)

- Most general: let's rename `UniLoRA*` to `UniLora` which makes it easier to remember how the method is typed in code (to be consistent with `LoraModel` and friends).
- I noticed that the copyright notice says 2024, let's use the correct starting year in all newly introduced files: 2025.
- Before pushing changes it is always good to run `make style` to correct any style issues automatically, otherwise the CI will not be happy

It is good to see that you've already added some tests. Let's extend those by adding UniLoRA to the `TEST_CASES` list in `tests/test_custom_models.py` - this will already give quite a bit of coverage. You can check the results by running:
```
pytest tests/test_custom_models.py
```

If those tests run we can extend the tests to `test_decoder_models.py` and `test_encoder_decoder_models.py` in a similar fashion.

For this to be mergable we also need documentation in `docs/source/package_reference/unilora.md` (added to the _toctree.md).

KaiyangLi1992 committed:
fix: update unilora implementation

KaiyangLi1992 committed:
remove unintended files from commit

KaiyangLi1992 committed:
Merge branch 'main' into unilora-submit

githubnemo commented on src/peft/tuners/unilora/config.py:
```diff
@@ -0,0 +1,153 @@
+# Copyright 2024-present the HuggingFace Inc. team.
```
```suggestion
# Copyright 2025-present the HuggingFace Inc. team.
```

githubnemo commented on src/peft/peft_model.py:
```diff
@@ -582,8 +582,14 @@ def from_pretrained(
+            if (
+                ("vblora_vector_bank" not in k and
+                "prompt_encoder" not in k) and
+                ("unilora_" not in k)
```
This disables the missing key check (i.e. when loading from a checkpoint and the checkpoint misses some keys the user won't be warned). Let's be a bit more precise about which keys are OK to be missing.

githubnemo commented on src/peft/tuners/unilora/model.py:
```diff
@@ -0,0 +1,196 @@
+        print(
+            f"Uni-LoRA params to-be-saved (float32-equivalent): {unilora_params:,d} "
+            f"|| total params to-be-saved: {(unilora_params + other_params):,d}"
+        )
```
This is still open (`get_nb...` is removed but `print_saveable_parameters` still uses it).

githubnemo commented on src/peft/tuners/unilora/model.py:
```diff
@@ -0,0 +1,210 @@
+    # (i.e., D) into `theta_d_length` buckets (i.e., d), where `d << D`. The resulting 
+    # index tensor determines which entry of the global UniLoRA vector `theta_d` 
+    # will be used to reconstruct each LoRA parameter.
+    def generate_index(self, lora_param_count, theta_d_length,proj_seed):
```
Let's make it a doc-string

```suggestion
    def generate_index(self, lora_param_count, theta_d_length,proj_seed):
        """Assigns a deterministic index in `[0, theta_d_length - 1]` to each LoRA parameter.
        This function maps the flattened LoRA parameter space of size `lora_param_count` 
        (i.e., D) into `theta_d_length` buckets (i.e., d), where `d << D`. The resulting 
        index tensor determines which entry of the global UniLoRA vector `theta_d` 
        will be used to reconstruct each LoRA parameter.
        """
```

githubnemo commented on src/peft/tuners/unilora/layer.py:
```diff
@@ -0,0 +1,293 @@
+            p = self.unilora_theta_d[adapter]
+            if p.device != device:
+                # Parameter migration: need .data to avoid creating new graph edges
+                self.unilora_theta_d[adapter].data = p.data.to(device)
```
I think `_ensure_device` can be replaced with `self._move_adapter_to_device_of_base_layer(adapter)` if all params are registered in `adapter_layer_names` or `other_param_names` in the layer class.

githubnemo commented on src/peft/utils/save_and_load.py:
```diff
@@ -183,6 +184,10 @@ def renamed_dora_weights(k):
+    elif config.peft_type == PeftType.UNILORA:
+        to_return = {}
+        to_return["base_model.unilora_theta_d." + adapter_name] = state_dict["base_model.unilora_theta_d." + adapter_name]
+      
```
Let's give the user the option to save the indices alongside the theta weights. This will allow the user to use a saved adapter even though the index sampling has changed for some reason.

VeRA (a few lines below this) can be used as inspiration.

githubnemo reviewed (commented):
Hey @KaiyangLi1992,

thanks for nicely integrating the requested changes. I've flagged some comments where it seemed that they've passed under the radar and added some new ones.

Please make sure that the branch is up-to-date with `main` and to run `make style` to run the code formatter.

Once all changes are integrated it would make sense to add an experiment for Uni-LoRA in the MetaMathQA test suite, maybe using `method_comparison/MetaMathQA/experiments/lora/*` as a starting point. Let's also add a quick example on how to use UniLoRA, e.g. using the `examples/miss_finetuning` as an example.

KaiyangLi1992 commented:
Hi @githubnemo , apologies again for the delay. I now have time to continue this PR.

I have updated the branch against the latest main, addressed the remaining review comments, and ran make style / make quality plus the UniLora-related tests. I also added the requested documentation/example/benchmark pieces.

Could you please reopen the PR when you have a chance? If you prefer a fresh PR instead, I can open a new one and reference this PR for review history. Thank you!

githubnemo commented:
Hi @KaiyangLi1992, unfortunately I can't reopen this PR, github doesn't let me - it seems you force-pushed or re-created the branch when the PR was already closed. The usual workaround for this is to reset your local branch to the commit that was last used in the PR (879e58d), push and then I can re-open the PR (as described [here](https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c)). This would let us keep the PR's history which would make reviewing easier.

Alternatively you can open a new PR with an updated description and link it here.

KaiyangLi1992 commented:
> Hi @KaiyangLi1992, unfortunately I can't reopen this PR, github doesn't let me - it seems you force-pushed or re-created the branch when the PR was already closed. The usual workaround for this is to reset your local branch to the commit that was last used in the PR ([879e58d](https://github.com/huggingface/peft/commit/879e58d5ebdacba1b47a7a56102b4ef7891bc283)), push and then I can re-open the PR (as described [here](https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c)). This would let us keep the PR's history which would make reviewing easier.
> 
> Alternatively you can open a new PR with an updated description and link it here.



Hi @githubnemo, thanks for the suggestion.

I opened a fresh PR here: https://github.com/huggingface/peft/pull/3257 .

The new PR is rebased on the latest `main` and includes the remaining review follow-ups from this PR, including the docs, tests, example, MetaMathQA config, deterministic index generation documentation, `init_weights`, and optional index saving.

Thanks again for the review!
